python/dgl/distributed/sparse_emb.py

```diff
@@ -0,0 +1,124 @@
+"""Define sparse embedding and optimizer."""
+
+from .. import backend as F
+
+def _get_ndata_name(name):
+    ''' This is to get the name of node data in the kvstore.
+
+    KVStore doesn't understand node data or edge data. We'll use a prefix to distinguish them.
+    '''
+    return 'node:' + name
+
+class SparseEmbedding:
+    ''' Sparse embeddings in the distributed KVStore.
+
+    The sparse embeddings are only used as node embeddings.
+
+    Parameters
+    ----------
+    g : DistGraph
+        The distributed graph object.
+    name : str
+        The name of the embeddings
+    shape : tuple of int
+        The shape of the embedding. The first dimension should be the number of nodes.
+    initializer : callable
+        The function to create the initial data.
+    '''
+    def __init__(self, g, name, shape, initializer):
+        assert shape[0] == g.number_of_nodes()
+        g._client.init_data(_get_ndata_name(name), shape, F.float32, 'node',
+                            g.get_partition_book(), initializer)
+        g._ndata._add(name)
+        g._node_embs.append(self)
+
+        self._tensor = g.ndata[name]
+        self._trace = []
+
+    def __call__(self, idx):
+        emb = F.attach_grad(self._tensor[idx])
+        self._trace.append((idx, emb))
+        return emb
+
+def sparse_adagrad_optimize(data_store, name, indices, data):
+    ''' Update the embeddings with sparse Adagrad.
+
+    This function runs on the KVStore server. It updates the gradients by scaling them
+    according to the state sum. The gradients have been adjusted by the learning rate
```

Comment:
Will this cause any problem? As the state_sum is calculated by the values adjusted by the learning rate. This is different from the SOTA impl.

Reply:
you are right. it should be adjusted by the learning rate beforehand.